orttraining/orttraining/python/training/ortmodule/_inference_manager.py

```diff
@@ -93,21 +98,24 @@ def forward(self, *inputs, **kwargs):
             # If creating the execution agent for the first time, this skip check will not take effect.
             # It will only take effect on subsequent forward calls.
             create_execution_session = False
-            if self._skip_check.is_set(_SkipCheck.SKIP_CHECK_EXECUTION_AGENT) == False or \
-                not self._execution_agent:
+            if self._skip_check.is_set(_SkipCheck.SKIP_CHECK_EXECUTION_AGENT) is False or \
+                    not self._execution_agent:
                 module_device = _utils.get_device_from_module(
                     self._original_module)
-                # The inference session should be created every time
-                # the graph was built or if the device changed between calls to forward
-                create_execution_session = build_graph or self._device != module_device
+
+                create_execution_session = (build_graph or self._device != module_device or
+                                            torch.are_deterministic_algorithms_enabled() is not
+                                            are_deterministic_algorithms_enabled())
```

Comment:
Do you think instead of keeping a global variable for determinism which could easily go out of consistency with what the session was created with, we could persist the `session_options` or more specifically `use_deterministic_compute` to compare `torch.are_deterministic_algorithms_enabled()` against?

Reply:
SessionOptions is a per model instance configuration, whereas determinism is a framework configuration, which is the same behavior as for PyTorch `are_deterministic_algorithms_enabled`

Reply:
I am ok with it being a framework level setting but just have concerns around it going out of sync with the torch's value and then causing unexpected results for the user.

Reply:
Got it, but they are part of the `create_execution_session` check for each `forward` call. When they mismatch, we update it internally and rebuild the execution session to reflect the change